docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,256 @@
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a developer I want the agent to be automatically upgraded to what the service
+expects, so that I can assume that the service will never proceed to install a
+cluster with a host that is running an incompatible version of the agent.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a comparison of the value of the
+`AGENT_DOCKER_IMAGE` environment variable passed to the service, and the
+`discovery_agent_version` header sent by the agent and stored in the `hosts`
+table.
+
+The `discovery_agent_version` currently reported by the agent is only the image
+tag. For example, if the full image reference is
+`registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141`
+the the agent will send only `v1.0.0-141` in the `discovery_agent_version`
+header. That will need to be changed in the agent, so that it sends the full
+image reference. The server will be changed to extract from that full image
+reference the tag where required, in particular for generation of metrics.
+
+In order to support both old and new agents the validation will not be a simple
+string comparison. Instead it will check if the `discovery_agent_version` sent
+by the agent is a full image reference or only a tag. If it is a full image
+reference then it will be a exact string comparision with the value of the
+`AGENT_DOCKER_IMAGE` environment variable. If it is only a tag, then it will be
+a comparison with the tag extracted from the value of the `AGENT_DOCKER_IMAGE`
+environment variable. In this last case a warning will be written to the log
+explaining that the comparision may be ureliable because only the tags were
+compared.
+
+To make this comparison useful it is important that the `AGENT_DOCKER_IMAGE`
+environment variable passed to the service doesn't use the `:latest` tag, or in
+general any moving tag, as that renders the comparison useless. In the _SaaS_
+stage environment that variable is already using a tag, so that shouldn't be a
+problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The service will check during startup the value of the `AGENT_DOCKER_IMAGE` and
+it will write a warning to the log if it is using the `:latest` tag.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
```

Comment:
Consider changing ⚠️ to ⌛ (or something less frightening) for this validation

Reply:
Done.